### ISSUE TITLE
Level-set surfaces in 3D: initialization performance troubles

### ISSUE THREAD

**Issue by thchr** [NanoComp/mpb#117]
Hi Steven,

This is closer to a feature-request or question than a bug-report.
I'm happy to try my hand at the implementation parts, but I'm not sure how best to approach it at this point.

I've been using MPB with level-set surfaces defined by the real part of a Fourier series in the reciprocal lattice vectors, like so:
```scm
; calculate the real part of the fourier sum
(define (calc-fourier-sum r)
    (fold-left + 0.0
        (map (lambda (g coef)
                (let ((g-dot-r (vector3-dot g r)))
                    (- (* (real-part coef) (cos g-dot-r))
                       (* (imag-part coef) (sin g-dot-r)))
                )
             ) gs coefs)
    )
)

; define associated level-set function (isolevel > 0 => inside; isolevel < 0 => outside)
(define (level-set r)
    (cond ((< (calc-fourier-sum r) isolevel) epsin)
          (else epsout))
)

; define material properties
(set! default-material (make material-function (epsilon-func level-set)))
```
where  `gs` is a list of `vector3`s (2π-scaled reciprocal lattice vectors), `coefs` is a list of `cnumbers`, and `isolevel`, `epsin`, and `epsout` are `numbers`.
This works, and I've used something like it without issue in 2D for a while. 

In 3D, however, evaluation time explodes: e.g. a 32x32x32 calculation will spend 20-30 minutes or more initializing epsilon (which makes me sad, because the subsequent MPB calculation I need takes ~1 min).

I'm assuming it takes such a long time because of the several layers of conversion back and forth between Guile and C.
I naively tried implementing the equivalent of `calc-fourier-sum` in C, with arguments `r`, `gs`, and `coefs` - but that is actually slower (and I don't get around the problem of needing to pass it back e.g. to `epsilon-func`); I assume because it actually just adds one more layer of conversion.

I'm wondering what is the right approach here: technically, I guess I could re-implement an equivalent of `epsilon-func` that is specialized to Fourier series - but I'm wondering if there is a better way? Would my life be easier if I did this in Python instead of Scheme?

**Comment by thchr** [comment]
Hi again @stevengj,

Looking at this again today, I was wondering whether you could confirm that the involved code paths are the ones below:
https://github.com/NanoComp/mpb/blob/0935b95c362873c2dd871646c5f12899e48af1c6/mpb/epsilon.c#L146-L159

Is the trouble that `gh_call1` is very slow or is the `material_input_type` call somehow also responsible?

I'm essentially seeing that using `epsilon-func` in 3D is extremely slow pretty much regardless of how complicated the Scheme function is (e.g., even for something simple like `(define (eps r) (+ 2 (vector3-x r)))`).

**Comment by stevengj** [comment]
Turn off subpixel averaging?

**Comment by stevengj** [comment]
The basic issue is that Scheme code is slow, and hence calling Scheme code for every pixel is slow.   It's worse if you do subpixel averaging, because then it evaluates many times per pixel.

**Comment by thchr** [comment]
I had thought turning off subpixel averaging would very significantly reduce the overall convergence? I'm guessing your point is that this could still be a worthwhile trade-off here. 
Are there any rules of thumb on how much subpixel averaging increases the overall `resolution`-derived sampling?

Is the only proper alternative to implement a basis-specific (e.g. Fourier) level set material class in C?

I guess there's also always the option of using a finely sampled .h5 file as input as well.

If you feel it is unfixable within the constraints of the .ctl DSL, feel free to close. Thanks!

**Comment by thchr** [comment]
For posterity's sake, I figured I update this in case anyone else stumbles across it: I managed to reduce the originally quoted initialization time by a factor of about 20-30 by a combination of a few things: 

- I was originally running this on Guile 2.0.x: switching over to Guile 3.0.4 brought an additional factor of about 4-5'ish. (v2.2.x brought a factor of 2'ish). Coincidentally, compatibility with Guile 3.0.4 seems fine.
- For a reason I don't quite understand, when including `.scm` or `.ctl` code in MPB, the resulting functions run slower than if they are included directly in Guile (curiously, the speed is restored if the methods are literally pasted into MPB's REPL). It's a bit awkward, but this can be bypassed by `(load ...)`ing relevant `.scm` files instead of `(include ...`) them, which forces compilation. This brings a factor of about 2-3.
- Misc improvements to the Guile implementation (e.g. using `vector` of `vector`s to store the data instead of `list`s). Another factor of 2-3.

Long story short: with some work, the Scheme side of things can be significantly improved. I think this issue could be closed.

**Comment by stevengj** [comment]
Okay.  

(Eventually it might be nice to support supplying a Fourier series directly to MPB, with the level set implemented in C.)
